examples/step-55/step-55.cc

```diff
@@ -0,0 +1,1212 @@
+    :
+    matrix (&m),
+    preconditioner (&preconditioner)
+  {}
+
+
+  // In Schur complement it is important that A^(-1) is accurate which is why we do it this way
+  template <class Matrix, class Preconditioner>
+  void InverseMatrix<Matrix,Preconditioner>::vmult (Vector<double>       &dst,
+                                                    const Vector<double> &src) const
+  {
+    SolverControl solver_control (src.size(), 1e-6*src.l2_norm());
+    SolverCG<>    cg (solver_control);
+
+    dst = 0;
+
+    cg.solve (*matrix, dst, src, *preconditioner);
+    count_mm = count_mm + solver_control.last_step();
+  }
+
+
+
+  template <class Preconditioner>
+  class SchurComplement : public Subscriptor
+  {
+  public:
+    SchurComplement (const BlockSparseMatrix<double> &system_matrix,
+                     const InverseMatrix<SparseMatrix<double>, Preconditioner> &A_inverse);
+
+    void vmult (Vector<double>       &dst,
+                const Vector<double> &src) const;
+
+  private:
+    const SmartPointer<const BlockSparseMatrix<double> > system_matrix;
+    const SmartPointer<const InverseMatrix<SparseMatrix<double>, Preconditioner> > A_inverse;
+
+    mutable Vector<double> tmp1, tmp2;
+  };
+
+
+
+  template <class Preconditioner>
+  SchurComplement<Preconditioner>::
+  SchurComplement (const BlockSparseMatrix<double> &system_matrix,
+                   const InverseMatrix<SparseMatrix<double>,Preconditioner> &A_inverse)
+    :
+    system_matrix (&system_matrix),
+    A_inverse (&A_inverse),
+    tmp1 (system_matrix.block(0,0).m()),
+    tmp2 (system_matrix.block(0,0).m())
+  {}
+
+
+  template <class Preconditioner>
+  void SchurComplement<Preconditioner>::vmult (Vector<double>       &dst,
+                                               const Vector<double> &src) const
+  {
+    // This is the transfer (understand)
+    system_matrix->block(0,1).vmult (tmp1, src);    // multiply with the top right block: B
+    A_inverse->vmult (tmp2, tmp1);                  // multiply with A^-1
+    system_matrix->block(1,0).vmult (dst, tmp2);    // multiply with the bottom left block: B^T
+  }
+
+  template <int dim>
+  StokesProblem<dim>::StokesProblem (const unsigned int degree)
+    :
+    degree (degree),
+    triangulation (Triangulation<dim>::maximum_smoothing),
+    fe (FE_Q<dim>(degree+1), dim, // Finite element for whole system
+        FE_Q<dim>(degree), 1),
+    velocity_fe (FE_Q<dim>(degree+1), dim), // Finite element for velocity-only
+    dof_handler (triangulation),
+    velocity_dof_handler (triangulation),
+    computing_timer (std::cout, TimerOutput::summary,
+                     TimerOutput::wall_times)
+  {}
+
+// @sect4{StokesProblem::setup_dofs}
+
+// This function sets up things differently based on if you want to use ILU or GMG as a preconditioner.
+  template <int dim>
+  void StokesProblem<dim>::setup_dofs ()
+  {
+    if (use_multigrid==false)
```

Comment:
a ton of code if duplicated in ``setup_dofs()`` and I think you are forgetting to clear out several objects when switching from MG to ILU or the other way around. Try to move everything that is common outside the if. You can also reorder some of the operations!